Added cost calculation for the construction calculator

pristit opened the pull request:
Added tooltip text detailing how many of each material is required.

I've edited the entries to only some of the actions in the skill resource json file, needs to be completed.

Possible to use the same actions and calculations to add a cost calculator for herblore.

![image](https://user-images.githubusercontent.com/17098942/119518648-6225b000-bd81-11eb-9f36-b0e849e447f3.png)

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java:
```diff
@@ -1,4 +1,4 @@
-/**
+/*
```
Revert this

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java:
```diff
@@ -50,14 +50,13 @@
-import net.runelite.client.plugins.skillcalculator.beans.SkillData;
-import net.runelite.client.plugins.skillcalculator.beans.SkillDataBonus;
-import net.runelite.client.plugins.skillcalculator.beans.SkillDataEntry;
+import net.runelite.client.plugins.skillcalculator.beans.*;
```
Revert this

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/beans/MaterialTuple.java:
```diff
@@ -0,0 +1,12 @@
+package net.runelite.client.plugins.skillcalculator.beans;
```
Needs license header

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/beans/MaterialTuple.java:
```diff
@@ -0,0 +1,12 @@
+import lombok.Getter;
+
+@Getter
+public class MaterialTuple
```
```suggestion
public class Material
```

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/beans/MaterialTuple.java:
```diff
@@ -0,0 +1,12 @@
+public class MaterialTuple
+{
+    private Integer id;
+    private String matName;
```
```suggestion
    private String name;
```

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java:
```diff
@@ -353,8 +351,33 @@ private void calculate()
+			if (finalCost == 0)
+				slot.setCost(null);
+			else
+				slot.setCost("Cost: " + QuantityFormatter.quantityToStackSize(finalCost) + " GP");
```
This can be in the `getMaterials()` null check block?

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java:
```diff
@@ -353,8 +351,33 @@ private void calculate()
+			if (materials.compareTo("") == 0)
+				slot.setToolTipText(null);
+			else
+				slot.setToolTipText("<html>" + materials + "</html>");
```
This can be in the `getMaterials()` null check block?

pristit commented:
Used the code style scheme provided in the wiki to format the files  I've touched (including the resource json file).

added the license header to Material.java, and added my name in the SkillCalculator.java credits (if this contribution isn't sufficient to add that let me know and I'll change it).

Changed the imports as was mentioned.

and I've moved the tooltip and cost setting into the getMaterials null check block (also added comments explaining them if it matters).

pristit commented:
After searching similar cases, it seems it's also related to issue #11154

Just needs to add similar field (albeit redundant) to the resource for prayer, as the code I've added only calculates cost if the skill resource file contains the "materials" field (to avoid trying to calculate cost for non buyable skills such as hunter/wood cutting)

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java:
```diff
@@ -353,8 +355,35 @@ private void calculate()
+				{
+					if (material != null)
+					{
+						actionCost += (itemManager.getItemPrice(material.getId()) * material.getAmount());
```
`getItemPrice` needs to be called on the client thread

pristit replied:
Injected client thread in the plugin and passed it via constructors.

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java:
```diff
@@ -353,8 +355,35 @@ private void calculate()
 				actionCount = (int) Math.ceil(neededXP / xp);
 			}
+			int actionCost = 0;
+			int finalCost;
```
This can be local in the if block

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java:
```diff
@@ -353,8 +355,35 @@ private void calculate()
 			{
 				actionCount = (int) Math.ceil(neededXP / xp);
 			}
+			int actionCost = 0;
```
This can be local in the if block

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java:
```diff
@@ -353,8 +355,35 @@ private void calculate()
 			}
+			int actionCost = 0;
+			int finalCost;
+			String materials = "";
```
This can be local in the if block

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/beans/Material.java:
```diff
@@ -0,0 +1,37 @@
+/*
```
```suggestion
/**
```

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/UIActionSlot.java:
```diff
@@ -128,8 +129,13 @@ public void mouseExited(MouseEvent mouseEvent)
 		uiLabelActions.setFont(FontManager.getRunescapeSmallFont());
 		uiLabelActions.setForeground(ColorScheme.LIGHT_GRAY_COLOR);
 
+		uiLabelCost = new JShadowedLabel("Unknown");
```
I think this should just default to empty/null and the null set block can be removed?
